File: TypeScript/Graphs/Dijkstra.ts
/**
 * Dijkstra's Algorithm Implementation
 * 
 * Time Complexity:
 * - O(V^2) with adjacency matrix.
 * - O(E log V) with adjacency list and priority queue.
 * 
 * Space Complexity: O(V) to store distances and visited state.
 */

export class Dijkstra {
    /**
     * Finds the shortest path from a start node to all other nodes.
     * @param graph Adjacency matrix representation of the graph
     * @param startNode The starting node index
     */
    static findShortestPath(graph: number[][], startNode: number): number[] {
        const n = graph.length;
        const distances = new Array(n).fill(Infinity);
        const visited = new Array(n).fill(false);

        distances[startNode] = 0;

        for (let i = 0; i < n - 1; i++) {
            const u = this.minDistance(distances, visited);
            visited[u] = true;

            for (let v = 0; v < n; v++) {
                if (!visited[v] && graph[u][v] !== 0 && distances[u] !== Infinity &&
                    distances[u] + graph[u][v] < distances[v]) {
                    distances[v] = distances[u] + graph[u][v];
                }
            }
        }

        return distances;
    }

    private static minDistance(distances: number[], visited: boolean[]): number {
        let min = Infinity;
        let minIndex = -1;

        for (let v = 0; v < distances.length; v++) {
            if (!visited[v] && distances[v] <= min) {
                min = distances[v];
                minIndex = v;
            }
        }

        return minIndex;
    }
}

// Example usage:
/*
const graph = [
    [0, 4, 0, 0, 0],
    [4, 0, 8, 0, 0],
    [0, 8, 0, 7, 0],
    [0, 0, 7, 0, 9],
    [0, 0, 0, 9, 0]
];
console.log(Dijkstra.findShortestPath(graph, 0));
*/

File: Python/Graphs/Dijkstra.py
"""
Dijkstra's Algorithm Implementation

Time Complexity: O(V^2) with adjacency matrix.
Space Complexity: O(V)
"""

import sys

def dijkstra(graph, start_node):
    n = len(graph)
    distances = [sys.maxsize] * n
    visited = [False] * n
    distances[start_node] = 0

    for _ in range(n):
        u = min_distance(distances, visited)
        visited[u] = True

        for v in range(n):
            if not visited[v] and graph[u][v] > 0 and \
               distances[u] != sys.maxsize and \
               distances[u] + graph[u][v] < distances[v]:
                distances[v] = distances[u] + graph[u][v]

    return distances

def min_distance(distances, visited):
    min_val = sys.maxsize
    min_index = -1

    for v in range(len(distances)):
        if distances[v] < min_val and not visited[v]:
            min_val = distances[v]
            min_index = v

    return min_index

# Example usage:
# if __name__ == "__main__":
#     graph = [
#         [0, 4, 0, 0, 0],
#         [4, 0, 8, 0, 0],
#         [0, 8, 0, 7, 0],
#         [0, 0, 7, 0, 9],
#         [0, 0, 0, 9, 0]
#     ]
#     print(dijkstra(graph, 0))
